File: 328.odd-even-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        heads = [
            ListNode(-1),  # Odd header
            ListNode(-1)   # Even header
        ]
        pointers = [
            heads[0],
            heads[1]
        ]
        idx = 0
        while (head):
            pointers[idx].next = head
            pointers[idx] = pointers[idx].next
            head = head.next
            idx = ((idx + 1) & 1)
        pointers[0].next = heads[1].next
        pointers[1].next = None
        return heads[0].next
            

File: 328.odd-even-linked-list/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun oddEvenList(head: ListNode?): ListNode? {
        if (head?.next == null) {
            return head
        }
        val heads = arrayOf(
            ListNode(-1),
            ListNode(-1)
        )
        val pointers = arrayOf(
            heads[0],
            heads[1]
        )

        var idx = 0
        var _h = head
        while (_h != null) {
            pointers[idx].next = _h
            pointers[idx] = pointers[idx].next
            _h = _h?.next
            idx = (idx + 1) and 1
        }
        pointers[0].next = heads[1].next
        pointers[1].next = null
        // this.printList(heads[0])
        // this.printList(heads[1])
        return heads[0].next
    }

    fun printList(head: ListNode?) {
        var _h = head
        val strBuffer = StringBuffer()
        while (_h != null) {
            strBuffer.append("${_h.`val`}->")
            _h = _h?.next
        }
        strBuffer.append("[null]")
        println(strBuffer)
    }
}
